### PR TITLE
Break up AccountsIndex lock

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
Holding AccountsIndex lock during scans blocks account stores

#### Summary of Changes
Builds on top of https://github.com/solana-labs/solana/pull/12126

Benchmark `bench_concurrent_scan_write` here shows 1000 inserts of new pubkeys into the storage taking about 9-10ms (80% still waiting on the AccountsIndex write lock), even as the index/accountsdb storage size grows. Before this change, the inserts tarting taking hundreds of ms.

Maybe we would use something like a https://github.com/crossbeam-rs/crossbeam/tree/master/crossbeam-skiplist to optimize contention the core lock farther, but there does not seem to be any stable alternatives yet.
Fixes #

### CODE REVIEW

File: runtime/Cargo.toml [hunk 1]
```diff
@@ -31,6 +32,7 @@ num_cpus = "1.13.0"
 rand = "0.7.0"
 rayon = "1.4.1"
 regex = "1.3.9"
+rental = "0.5.4"
```
Comment (ryoqun) [inline runtime/Cargo.toml:35]: :eyes: https://crates.io/crates/rental

File: runtime/benches/accounts.rs [hunk 2]
```diff
@@ -141,3 +142,103 @@ fn bench_delete_dependencies(bencher: &mut Bencher) {
         accounts.accounts_db.clean_accounts(None);
     });
 }
+
+#[bench]
+#[ignore]
+fn bench_concurrent_read_write(bencher: &mut Bencher) {
+    let num_readers = 5;
+    let accounts = Arc::new(Accounts::new(
+        vec![PathBuf::from("concurrent_read_write")],
```
Comment (ryoqun) [inline runtime/benches/accounts.rs:151]: nits recognize FARF_DIR?

File: runtime/benches/accounts.rs [hunk 3]
```diff
@@ -141,3 +142,103 @@ fn bench_delete_dependencies(bencher: &mut Bencher) {
...
+            accounts.store_slow(slot + 1, &Pubkey::new_rand(), &account);
+        }
+    })
+}
+
+#[bench]
+fn bench_concurrent_scan_write(bencher: &mut Bencher) {
+    let num_readers = 5;
+    let accounts = Arc::new(Accounts::new(
+        vec![PathBuf::from("concurrent_read_write")],
```
Comment (ryoqun) [inline runtime/benches/accounts.rs:201]: nits recognize FARF_DIR?

File: runtime/benches/accounts.rs [hunk 4]
```diff
@@ -141,3 +142,103 @@ fn bench_delete_dependencies(bencher: &mut Bencher) {
...
+    for _ in 0..num_readers {
+        let accounts = accounts.clone();
+        let pubkeys = pubkeys.clone();
+        Builder::new()
+            .name("readers".to_string())
+            .spawn(move || {
+                let mut rng = rand::thread_rng();
+                loop {
+                    let i = rng.gen_range(0, num_keys);
+                    accounts.load_slow(&HashMap::new(), &pubkeys[i]);
```
Comment (ryoqun) [inline runtime/benches/accounts.rs:177]: nits: `unwrap` should be safe, right?

nits: also how about using `test::black_box` in the fear of human-level clever `rustc` in the distant future? ;)

File: runtime/benches/accounts.rs [hunk 5]
```diff
@@ -141,3 +142,103 @@ fn bench_delete_dependencies(bencher: &mut Bencher) {
...
+    ));
+    let num_keys = 1000;
+    let slot = 0;
+    accounts.add_root(slot);
+    let default_owner = Account::default().owner;
+    let pubkeys: Arc<Vec<_>> = Arc::new(
+        (0..num_keys)
+            .map(|_| {
+                let pubkey = Pubkey::new_rand();
+                let account = Account::new(1, 0, &Account::default().owner);
```
Comment (ryoqun) [inline runtime/benches/accounts.rs:212]: nits: can use `default_owner`?

File: runtime/benches/accounts.rs [hunk 6]
```diff
@@ -141,3 +142,103 @@ fn bench_delete_dependencies(bencher: &mut Bencher) {
...
+    );
+
+    for _ in 0..num_readers {
+        let accounts = accounts.clone();
+        let pubkeys = pubkeys.clone();
+        Builder::new()
+            .name("readers".to_string())
+            .spawn(move || {
+                loop {
+                    accounts.load_by_program(&HashMap::new(), &default_owner);
```
Comment (ryoqun) [inline runtime/benches/accounts.rs:226]: nits: like https://github.com/solana-labs/solana/pull/12787/files#r503695336, how about `assert!(!loaded_accounts.is_empty())` and `test::black_box()`?

File: runtime/benches/accounts.rs [hunk 7]
```diff
@@ -141,3 +142,103 @@ fn bench_delete_dependencies(bencher: &mut Bencher) {
...
+
+    let num_new_keys = 1000;
+    let new_accounts: Vec<_> = (0..num_new_keys)
+        .map(|_| Account::new(1, 0, &Account::default().owner))
+        .collect();
+    bencher.iter(|| {
+        for account in &new_accounts {
+            // Write to a different slot than the one being read from. Because
+            // there's a new account pubkey being written to every time, will
+            // compete for the accounts index lock on every store
```
Comment (ryoqun) [inline runtime/benches/accounts.rs:191]: good comment!

File: runtime/benches/accounts.rs [hunk 8]
```diff
@@ -141,3 +142,103 @@ fn bench_delete_dependencies(bencher: &mut Bencher) {
...
+            accounts.store_slow(slot + 1, &Pubkey::new_rand(), &account);
+        }
```
Comment (ryoqun) [inline runtime/benches/accounts.rs:193]: nits: how about extracting to something like this and share with `bench_concurrent_scan_write`, too?: `store_accounts_with_possible_contention`. (`possible` is explicitly picked in the consideration of the improvement in this pr itself). :)

File: runtime/benches/accounts.rs (hunk 4 above)
Comment (carllin) [reply runtime/benches/accounts.rs:177]: good suggestion, updated on other PR as well!

File: runtime/src/accounts.rs [hunk 9]
```diff
@@ -302,8 +302,7 @@ impl Accounts {
     ) -> Vec<(Result<TransactionLoadResult>, Option<HashAgeKind>)> {
         //PERF: hold the lock to scan for the references, but not to clone the accounts
         //TODO: two locks usually leads to deadlocks, should this be one structure?
```
Comment (ryoqun) [inline runtime/src/accounts.rs:304]: Grab your fruit: now prime time to remove this outdated comment :)

File: runtime/src/accounts_index.rs [hunk 10]
```diff
@@ -1,38 +1,276 @@
...
 use std::{
     collections::{BTreeMap, HashMap, HashSet},
     ops::RangeBounds,
-    sync::{RwLock, RwLockReadGuard},
+    sync::{Arc, RwLock},
 };
 
+const ITER_BATCH_SIZE: usize = 1000;
+
+rental! {
```
Comment (ryoqun) [inline runtime/src/accounts_index.rs:15]: interesting. :)

File: runtime/src/accounts_index.rs [hunk 11]
```diff
@@ -1,38 +1,276 @@
...
+            .read()
+            .unwrap()
+            .range((self.start_bound, self.end_bound))
+            .map(|(pubkey, account_map_entry)| {
+                (*pubkey, account_map_entry.clone())
+            })
+            .take(ITER_BATCH_SIZE)
+            .collect();
+
+        if chunk.is_empty() {
```
Comment (ryoqun) [inline runtime/src/accounts_index.rs:185]: nits: there is very and very small corner case of violating `iter` invariant: we could return `Some` even after we returned `None` here if we got newer Pubkey inserted, which would be compared after current `self.start_bound`. how about closing this hole now? Consider data parallelism from `rayon` and imagine how hard to spot this bug in the future from insane behavior. :)

Sadly, (start, end) = (unbound, unbound) or anything like that wouldn't work to mark the state. So, we must add another flag `finished` or like that and set it to `true` here.

File: runtime/src/accounts_index.rs [hunk 12]
```diff
@@ -1,38 +1,276 @@
...
 
 pub type RefCount = u64;
-type AccountMapEntry<T> = (AtomicU64, RwLock<SlotList<T>>);
 pub type AccountMap<K, V> = BTreeMap<K, V>;
 
+#[derive(Clone, Debug)]
+pub struct AccountMapEntry<T> {
+    ref_count: Arc<AtomicU64>,
+    pub slot_list: Arc<RwLock<SlotList<T>>>,
+}
```
Comment (ryoqun) [inline runtime/src/accounts_index.rs:52]: I wonder we could reduce Arc to 1 from 2 by line this, it's a bit hideous but maybe worth effort? (we should be keen the size of AcccontIndex memory footprint (Arc costs 2-word memory for each, so we're increasing entry size by `4 * usize` here per `Account`.) and reduced general update cost from updating Arc twice everytime accessed).

```rust
pub struct AccountMapEntryInner<T> {
    ref_count: AtomicU64,
    pub slot_list: RwLock<SlotList<T>>,
}
type AccountMapEntry<T> = Arc<AccountMapEntryInner>;
```

**Comment by ryoqun** [comment]
> Benchmark `bench_concurrent_scan_write` here shows 1000 inserts of new pubkeys into the storage taking about 9-10ms (80% still waiting on the AccountsIndex write lock), even as the index/accountsdb storage size grows. Before this change, the inserts tarting taking hundreds of ms.

I'm really happy to see this report. :)


How about any perf changes of `colo-gpu-perf-4-val-1-client.yml` with this(#12787) & previous dashmap (#12126),  compared to the base commit?

File: runtime/benches/accounts.rs (hunk 8 above)
Comment (carllin) [reply runtime/benches/accounts.rs]: Updated!

File: runtime/src/accounts_index.rs (hunk 11 above)
Comment (carllin) [reply runtime/src/accounts_index.rs:189]: Fixed and added test!

File: runtime/benches/accounts.rs [hunk 13]
```diff
@@ -196,3 +193,31 @@ fn bench_concurrent_read_write(bencher: &mut Bencher) {
...
+}
+
+#[bench]
+#[ignore]
+fn bench_concurrent_scan_write(bencher: &mut Bencher) {
+    store_accounts_with_possible_contention(
+        "concurrent_scan_write",
+        bencher,
+        |accounts, pubkeys| loop {
+            accounts.load_by_program(&HashMap::new(), &Account::default().owner);
```
Comment (ryoqun) [inline runtime/benches/accounts.rs:220]: nits: test::black_box here as well?

File: runtime/src/accounts_index.rs [hunk 14]
```diff
@@ -160,72 +405,31 @@ impl<'a, T: 'a + Clone> AccountsIndex<T> {
...
-
-    // Try to update an item in account_maps. If the account is not
-    // already present, then the function will return back Some(account_info) which
-    // the caller can then take the write lock and do an 'insert' with the item.
-    // It returns None if the item is already present and thus successfully updated.
-    pub fn update(
+    // Updates the given pubkey at the given slot with the new account information.
+    // Returns true if the pubkey was newly inserted into the index, otherwise, if the
+    // pubkey updates an existing entry in the index, returns false.
+    pub fn update_or_create_if_missing(
```
Comment (ryoqun) [inline runtime/src/accounts_index.rs:411]: nits: I welcome less confusing name so thanks for taking a bite on this. but this is a bit long relatively for frequently called method. how about just `upsert`?

File: runtime/src/accounts_index.rs (hunk 10 above)
Comment (ryoqun) [reply runtime/src/accounts_index.rs:15]: thinking on this.... hmm.

Comment (carllin) [reply runtime/src/accounts_index.rs]: Updated with the ouroboros suggestion, thanks!

File: runtime/Cargo.toml (hunk 1 above)
Comment (carllin) [reply runtime/Cargo.toml]: gone!

File: runtime/src/accounts_index.rs (hunk 14 above)
Comment (carllin) [reply runtime/src/accounts_index.rs:436]: that works for me!

File: runtime/benches/accounts.rs (hunk 13 above)
Comment (carllin) [reply runtime/benches/accounts.rs:223]: ah forgot, thanks!

File: core/src/validator.rs [hunk 15]
```diff
@@ -1056,7 +1056,6 @@ fn report_target_features() {
         // Validator binaries built on a machine with AVX support will generate invalid opcodes
         // when run on machines without AVX causing a non-obvious process abort.  Instead detect
         // the mismatch and error cleanly.
-        #[target_feature(enable = "avx")]
```
Comment (ryoqun) [inline core/src/validator.rs:1059]: nits: unrelated change?

Comment (carllin) [reply core/src/validator.rs:1059]: uh super annoying, my `cargo fmt` keeps deleting this line... thanks!

File: runtime/src/accounts_db.rs [hunk 16]
```diff
@@ -1311,37 +1303,52 @@ impl AccountsDB {
         accounts_index: &AccountsIndex<AccountInfo>,
         pubkey: &Pubkey,
     ) -> Option<(Account, Slot)> {
-        let (lock, index) = accounts_index.get(pubkey, Some(ancestors), None)?;
-        let slot = lock[index].0;
+        let (slot, store_id, offset) = {
+            let (lock, index) = accounts_index.get(pubkey, Some(ancestors), None)?;
+            let slot_list = lock.slot_list();
+            let slot = slot_list[index].0;
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:1309]: nits: can't we do like this destructuring?

```rust
let (slot, {store_id, offset, ...}) = slot_list[index];
```

File: runtime/src/accounts_db.rs [hunk 17]
```diff
@@ -1311,37 +1303,52 @@ impl AccountsDB {
...
-        let info = &lock[index].1;
+        let (slot, store_id, offset) = {
+            let (lock, index) = self
+                .accounts_index
+                .get(pubkey, Some(ancestors), None)
+                .unwrap();
+            let slot_list = lock.slot_list();
+            let slot = slot_list[index].0;
+            let store_id = slot_list[index].1.store_id;
+            let offset = slot_list[index].1.offset;
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:1337]: same as this: https://github.com/solana-labs/solana/pull/12787/files?diff=split&w=1#r505171426

File: core/src/validator.rs (hunk 15 above)
Comment (ryoqun) [reply core/src/validator.rs]: protip: `(unset RUSTFLAGS && cargo +1.45.0 fmt --all)` This way I can live with ever-evolving nightly. :)

File: runtime/src/accounts_index.rs [hunk 18]
```diff
@@ -1,38 +1,305 @@
...
-    sync::{RwLock, RwLockReadGuard},
+    ops::{Deref, DerefMut, Range, RangeBounds},
+    sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard},
 };
+const ITER_BATCH_SIZE: usize = 1000;
+
+#[self_referencing]
+pub struct OwningReadGuard<T>
+where
+    T: 'static,
```
Comment (ryoqun) [inline runtime/src/accounts_index.rs:18]: nits: no need `where`?:

```rust
pub struct OwningReadGuard<T: 'static>
{
   ...
```

File: runtime/src/accounts_index.rs [hunk 19]
```diff
@@ -1,38 +1,305 @@
...
+    type Target = T;
+    fn deref(&self) -> &Self::Target {
+        &*self.guard
+    }
+}
+
+#[self_referencing]
+pub struct OwningWriteGuard<T>
+where
+    T: 'static,
```
Comment (ryoqun) [inline runtime/src/accounts_index.rs:35]: same as https://github.com/solana-labs/solana/pull/12787/files?diff=split&w=1#r505174649

File: runtime/src/accounts_db.rs (hunk 16 above)
Comment (carllin) [reply runtime/src/accounts_db.rs:1309]: yup totally!

Comment (carllin) [reply runtime/src/accounts_db.rs]: yup totally, updated!

**Comment by carllin** [comment]
@ryoqun been running for past 20 hours or so with snapshot verification: https://metrics.solana.com:3000/d/monitor-edge/cluster-telemetry-edge?orgId=2&from=1602723051575&to=1602796347486&panelId=35&fullscreen&var-datasource=Solana%20Metrics%20(read-only)&var-testnet=mainnet-beta&var-hostid=aeypRbHck9k5zGuftrdSaUMfjJyTrjtaGJEtALt835y, no issues yet!

File: runtime/src/accounts_index.rs (hunk 12 above)
Comment (carllin) [reply runtime/src/accounts_index.rs:72]: @ryoqun  this is doable, but I think it makes the self-referential struct a little uglier because now you have to wrap the entire `AccountsMapEntryInner` instead of just the lock inside the `self-reference` tag. Here's an example:  https://github.com/solana-labs/solana/pull/12787/commits/50a9b2647d6a47c0758b78e277731e5731accd5c

What do you think?

Comment (carllin) [reply runtime/src/accounts_index.rs]: @ryoqun this is much cleaner with the ouroboros crate here: https://github.com/solana-labs/solana/pull/12787/commits/970b2c78aa1a04a980500591fd2cc5cbc6af6be5

File: runtime/src/accounts_index.rs [hunk 20]
```diff
@@ -1,38 +1,297 @@
...
+    }
+
+    // Try to update an item in account_maps. If the account is not
+    // already present, then the function will return back Some(account_info) which
+    // the caller can then take the write lock and do an 'insert' with the item.
+    // It returns None if the item is already present and thus successfully updated.
+    pub fn update(&mut self, slot: Slot, account_info: T, reclaims: &mut SlotList<T>) {
+        // filter out other dirty entries from the same slot
+        let mut same_slot_previous_updates: Vec<(usize, (Slot, &T))> = self
+            .slot_list_mut()
```
Comment (ryoqun) [inline runtime/src/accounts_index.rs:120]: nits: doesn't need to be `_mut`? So define `.slot_list()` and use it?

File: runtime/src/accounts_index.rs [hunk 21]
```diff
@@ -1,38 +1,297 @@
...
+            .iter()
+            .enumerate()
+            .filter_map(|(i, (s, value))| {
+                if *s == slot {
+                    Some((i, (*s, value)))
```
Comment (ryoqun) [inline runtime/src/accounts_index.rs:125]: nits: this looks so similar to the enclosing closure signature. Can we use `filter` instead of `filter_map` somehow? Because this is fairly in the tight loop. I want to squeeze out anything unneeded as much as possible. :)

File: runtime/src/accounts_index.rs [hunk 22]
```diff
@@ -1,38 +1,297 @@
...
+    }
+
+    pub fn ref_count(&self) -> &AtomicU64 {
+        &self.owned_entry.ref_count
+    }
+
+    // Try to update an item in account_maps. If the account is not
+    // already present, then the function will return back Some(account_info) which
+    // the caller can then take the write lock and do an 'insert' with the item.
+    // It returns None if the item is already present and thus successfully updated.
```
Comment (ryoqun) [inline runtime/src/accounts_index.rs:116]: outdated comment?

**Comment by ryoqun** [comment]
> @ryoqun been running for past 20 hours or so with snapshot verification: https://metrics.solana.com:3000/d/monitor-edge/cluster-telemetry-edge?orgId=2&from=1602723051575&to=1602796347486&panelId=35&fullscreen&var-datasource=Solana%20Metrics%20(read-only)&var-testnet=mainnet-beta&var-hostid=aeypRbHck9k5zGuftrdSaUMfjJyTrjtaGJEtALt835y, no issues yet!

cool. :)

also, had the validator been slummed by `getProgramAccounts` RPC reqs? and no slowdowns of replay and no odd `panic!`s?

> How about any perf changes of `colo-gpu-perf-4-val-1-client.yml` with this(#12787) & previous dashmap (#12126), compared to the base commit?

As a ground wrapping-up of past various AccountsDB related improvements by you, how about writing like this? Too much hustle?  This is done for AccountIndex's `HashMap` => `BTreeMap` change.:
>  https://github.com/solana-labs/solana/pull/9527#issuecomment-627210347

you can peek around this build jobs to know how to create these pdfs or ask me :) : https://buildkite.com/solana-labs/system-performance-tests/builds?branch=pull%2F9527%2Fhead

also, please take a look at my previous similar reports for inspiration:

> https://github.com/solana-labs/solana/pull/8436#issuecomment-591865918
> https://github.com/solana-labs/solana/pull/8436#issuecomment-592390251

File: runtime/src/accounts_index.rs (hunk 21 above)
Comment (carllin) [reply runtime/src/accounts_index.rs:125]: yeah for sure, removed!

File: runtime/src/accounts_index.rs [hunk 23]
```diff
@@ -1,38 +1,297 @@
...
+            ));
+        }
+
+        (w_account_entry.unwrap(), is_newly_inserted)
+    }
+
+    pub fn handle_dead_keys(&self, dead_keys: Vec<Pubkey>) {
+        if !dead_keys.is_empty() {
+            for key in &dead_keys {
+                let mut w_index = self.account_maps.write().unwrap();
```
Comment (ryoqun) [inline runtime/src/accounts_index.rs:289]: just making sure: `w_index` is moved inside loop body to mitigate longer write lock?

File: runtime/src/accounts_index.rs [hunk 24]
```diff
@@ -1,38 +1,297 @@
...
+            });
+            w_account_entry = Some(WriteAccountMapEntry::from_account_map_entry(
+                account_entry.clone(),
+            ));
+        }
+
+        (w_account_entry.unwrap(), is_newly_inserted)
+    }
+
+    pub fn handle_dead_keys(&self, dead_keys: Vec<Pubkey>) {
```
Comment (ryoqun) [inline runtime/src/accounts_index.rs:286]: nits: how about taking a `&[Pubkey]`? I donno why clippy didn't warn for this. ;)

File: runtime/src/accounts_index.rs [hunk 25]
```diff
@@ -1,38 +1,297 @@
...
+        if !dead_keys.is_empty() {
+            for key in &dead_keys {
+                let mut w_index = self.account_maps.write().unwrap();
+                if let Some(account_entry) = w_index.get(key) {
+                    if account_entry.slot_list.read().unwrap().is_empty() {
+                        w_index.remove(key);
+                    }
```
Comment (ryoqun) [inline runtime/src/accounts_index.rs:293]: nits: I think we could reduce the lookup from 2 times to once like this:

```rust
if let Some(index_entry) = w_index.entry(key) {
    if (...) {
        index_entry.remove();
    }
}
```

**Comment by ryoqun** [comment]
2nd review pass is done. I think this is really reaching to the lgtm. Thanks for defending my wave of nits attacks patiently. Well, I'm very tend to be interrupted by other concurrent tasks..

File: runtime/src/accounts_index.rs (hunk 23 above)
Comment (carllin) [reply runtime/src/accounts_index.rs:289]: hmm yeah, it should be a pretty fast loop though, so maybe it's not necessary?

File: runtime/src/accounts_index.rs (hunk 25 above)
Comment (carllin) [reply runtime/src/accounts_index.rs:293]: slight variation on this using `Occupied` instead of `Some`, but updated!

File: runtime/src/accounts_index.rs [hunk 26]
```diff
@@ -1,38 +1,297 @@
...
+    fn deref(&self) -> &Self::Target {
+        &*self.guard
+    }
+}
+
+impl<T> DerefMut for OwningWriteGuard<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut *self.guard
+    }
+}
```
[the remark this answers is not in the document]
Comment (ryoqun) [reply runtime/src/accounts_index.rs:50]: maybe, my joke was so subtle... Well, you can remove this 30+ lines because these aren't used anymore at all. :)

File: runtime/src/accounts_index.rs (hunk 23 above)
Comment (ryoqun) [reply runtime/src/accounts_index.rs:288]: hmm, I'm not sure too for the perf. difference.. Anyway, I just wanted to confirm this was somewhat intentional or just typo.

**Comment by carllin** [comment]
@ryoqun, I couldn't get the PDF's to scale properly, but I was able to collect some metrics:

With the DashMap + Index change: https://metrics.solana.com:3000/d/monitor-edge/cluster-telemetry-edge?orgId=2&from=1602901242387&to=1602902152277&panelId=35&var-datasource=Solana%20Metrics%20(read-only)&var-testnet=testnet-dev-carl&var-hostid=All

Without DashMap + Index change: https://metrics.solana.com:3000/d/monitor-edge/cluster-telemetry-edge?orgId=2&from=1602902319061&to=1602903305200&panelId=35&var-datasource=Solana%20Metrics%20(read-only)&var-testnet=testnet-dev-carl&var-hostid=All

The confirmation + TPS are about the same, the biggest difference seems to be replay has fewer spikes with the changes: https://metrics.solana.com:3000/d/monitor-edge/cluster-telemetry-edge?orgId=2&from=1602901207646&to=1602903332757&panelId=35&fullscreen&var-datasource=Solana%20Metrics%20(read-only)&var-testnet=testnet-dev-carl&var-hostid=All&refresh=1m

**Comment by ryoqun** [comment]
> @ryoqun, I couldn't get the PDF's to scale properly

Thanks for running these perf results, anyway! I created them for future reference.

[break-up-accounts-index-before.pdf](https://github.com/solana-labs/solana/files/5397459/break-up-accounts-index-before.pdf)
[break-up-accounts-index-after.pdf](https://github.com/solana-labs/solana/files/5397458/break-up-accounts-index-after.pdf)

I'm still looking the result. But, in general, there should be no bad perf. degradation and spike reduced a lot for some reason. Maybe the root contributor to it is the reduction of squashing, which I didn't expected but a good outcome. :)

File: runtime/src/accounts_db.rs [hunk 27]
```diff
@@ -725,28 +723,37 @@ impl AccountsDB {
...
+                                *pubkey,
+                                self.accounts_index.roots_and_ref_count(&locked_entry),
+                            );
                         }
-                        if accounts_index.uncleaned_roots.contains(slot) {
-                            // Assertion enforced by `accounts_index.get()`, the latest slot
-                            // will not be greater than the given `max_clean_root`
-                            if let Some(max_clean_root) = max_clean_root {
-                                assert!(*slot <= max_clean_root);
-                            }
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:748]: nits: this assert! is no longer needed?

**Review by ryoqun — APPROVED** [review]
LGTM with nits!

Thanks for working so hard on AccountsDB.

File: runtime/src/accounts_db.rs (hunk 27 above)
Comment (carllin) [reply runtime/src/accounts_db.rs:748]: oh good catch, readded!
